File: src/test/java/nihilite/javaagentClasspathDriver.java
package nihilite;

import clojure.java.api.Clojure;
import clojure.lang.IFn;

public final class javaagentClasspathDriver {

    public static volatile int PASS = 0;
    public static volatile int FAIL = 0;

    private javaagentClasspathDriver() {}

    public static void main(String[] args) throws Exception {
        if (args.length > 0 && "spawn-jar-smoke".equals(args[0])) {
            spawnJavaJarSmoke(args);
            return;
        }
        runNreplMiscProbe();
        runRequireNreplServerProbe();
        runWorkerEquivalentProbe();
        runBootThenStartServerProbe();
        if (FAIL == 0 && PASS == 4) {
            System.out.println("DRIVER_PASS javaagent classpath: 4/4 probes OK");
            System.exit(0);
        } else {
            System.err.println("DRIVER_FAIL javaagent classpath: pass=" + PASS + " fail=" + FAIL);
            System.exit(1);
        }
    }

    private static void spawnJavaJarSmoke(String[] argv) throws Exception {
        String nihiliteJar = argv[1];
        String initScript = argv.length > 2 ? argv[2] : "";
        String initForm = initScript.isEmpty()
                ? "(do (require 'clojure.repl) (in-ns 'user))"
                : "(load-file \"" + initScript.replace("\\", "\\\\").replace("\"", "\\\"") + "\")";
        String[] cmd = new String[] {
                System.getProperty("java.home") + "/bin/java",
                "-Djdk.attach.allowAttachSelf=true",
                "-XX:+EnableDynamicAgentLoading",
                "-javaagent:" + nihiliteJar,
                "-Dnihilite.init=" + initForm,
                "-Dnihilite.port=0",
                "-Dnihilite.bind=127.0.0.1",
                "-jar", nihiliteJar
            };
        Process p = new ProcessBuilder(cmd).redirectErrorStream(true).start();
        java.io.InputStream is = p.getInputStream();
        java.io.ByteArrayOutputStream pipe = new java.io.ByteArrayOutputStream();
        byte[] buf = new byte[4096];
        Thread reader = new Thread(() -> {
            try {
                int n;
                while ((n = is.read(buf)) != -1) {
                    synchronized (pipe) { pipe.write(buf, 0, n); }
                }
            } catch (Throwable ignored) {}
        });
        reader.setDaemon(true);
        reader.start();

        boolean bound = false;
        boolean initDone = false;
        long deadline = System.nanoTime() + java.util.concurrent.TimeUnit.SECONDS.toNanos(45);
        while (System.nanoTime() < deadline && p.isAlive()) {
            Thread.sleep(200);
            String snap;
            synchronized (pipe) { snap = pipe.toString(); }
            if (!initDone && snap.contains("init eval done")) initDone = true;
            if (!bound && snap.contains("nREPL bencode clients may connect")) bound = true;
            if (initDone && bound) break;
        }
        p.destroyForcibly();
        try { p.waitFor(5, java.util.concurrent.TimeUnit.SECONDS); } catch (InterruptedException ie) { Thread.currentThread().interrupt(); }
        String log;
        synchronized (pipe) { log = pipe.toString(); }
        if (!bound || !initDone) {
            System.err.println("javaagentClasspathDriver: jar-smoke incomplete (bound=" + bound
                    + ", initDone=" + initDone + ")");
            System.err.println("...captured log (full):\n" + log);
            FAIL++;
            return;
        }
        System.out.println("javaagentClasspathDriver: jar-smoke (nrepl server bound + init ran) OK");
        PASS++;
    }

    private static void runNreplMiscProbe() {
        try {
            ClassLoader sys = ClassLoader.getSystemClassLoader();
            System.out.println("javaagentClasspathDriver: system loader class = " + sys.getClass().getName());
            System.out.println("javaagentClasspathDriver: system loader is URLClassLoader? "
                    + (sys instanceof java.net.URLClassLoader));
            System.out.println("javaagentClasspathDriver: nrepl/misc.clj via system = "
                    + sys.getResource("nrepl/misc.clj"));
            IFn require = Clojure.var("clojure.core", "require");
            require.invoke(Clojure.read("nrepl.misc"));
            System.out.println("javaagentClasspathDriver: require nrepl.misc OK");
            PASS++;
        } catch (Throwable t) {
            System.err.println("javaagentClasspathDriver: require nrepl.misc FAILED");
            logCauseChain(t);
            FAIL++;
        }
    }

    private static void runRequireNreplServerProbe() {
        try {
            IFn require = Clojure.var("clojure.core", "require");
            require.invoke(Clojure.read("nrepl.server"));
            System.out.println("javaagentClasspathDriver: require nrepl.server OK");
            PASS++;
        } catch (Throwable t) {
            System.err.println("javaagentClasspathDriver: require nrepl.server FAILED");
            logCauseChain(t);
            FAIL++;
        }
    }

    private static void runWorkerEquivalentProbe() throws Exception {
        IFn require = Clojure.var("clojure.core", "require");
        final Throwable[] workerError = { null };
        final boolean[] workerDone = { false };
        Thread workerThread = new Thread(new Runnable() {
            @Override public void run() {
                try {
                    require.invoke(Clojure.read("nihilite.registry"));
                    require.invoke(Clojure.read("nihilite.boot"));
                } catch (Throwable t) {
                    synchronized (workerError) { workerError[0] = t; }
                } finally {
                    synchronized (workerDone) { workerDone[0] = true; workerDone.notifyAll(); }
                }
            }
        }, "driver-worker-equivalent");
        workerThread.setDaemon(true);
        workerThread.start();
        try {
            require.invoke(Clojure.read("nihilite.boot"));
        } catch (Throwable t) {
            logCauseChain(t);
        }
        synchronized (workerDone) {
            while (!workerDone[0]) workerDone.wait();
        }
        Throwable we;
        synchronized (workerError) { we = workerError[0]; }
        if (we != null) {
            System.err.println("javaagentClasspathDriver: worker-equivalent concurrent require FAILED");
            logCauseChain(we);
            FAIL++;
        } else {
            System.out.println("javaagentClasspathDriver: worker-equivalent concurrent require OK");
            PASS++;
        }
    }

    private static void logCauseChain(Throwable t) {
        Throwable cause = t;
        while (cause != null) {
            System.err.println("  " + cause.getClass().getName()
                    + (cause.getMessage() == null ? "" : ": " + cause.getMessage()));
            cause = cause.getCause();
        }
    }

    private static void runBootThenStartServerProbe() {
        IFn require = Clojure.var("clojure.core", "require");
        IFn startFn = Clojure.var("nihilite.boot", "start!");
        IFn stopFn  = Clojure.var("nihilite.boot", "stop!");
        try {
            require.invoke(Clojure.read("nihilite.boot"));
            Object handle = startFn.invoke(Clojure.read("{:port 0 :bind \"127.0.0.1\"}"));
            try {
                System.out.println("javaagentClasspathDriver: nihilite.boot/start! OK");
                PASS++;
            } finally {
                try { stopFn.invoke(handle); } catch (Throwable t) { /* ignore */ }
            }
        } catch (Throwable t) {
            System.err.println("javaagentClasspathDriver: nihilite.boot/start! FAILED");
            logCauseChain(t);
            FAIL++;
        }
    }
}

File: src/test/clojure/nihilite/javaagent_classpath_driver.clj
(ns nihilite.javaagent-classpath-driver
  "Replaces src/test/java/nihilite/javaagentClasspathDriver.java.

   Runs four probes against the agent classpath, and a fifth jar-smoke
   path that spawns `java -javaagent:<jar>` and verifies the embedded
   nREPL server boots within the timeout.

   Invoked from build.clj as `java nihilite.javaagentClasspathDriver`."
  (:require [nihilite.boot :as boot]
            [clojure.string :as str]
            [nrepl.server :as nrepl.server])
  (:import [java.io ByteArrayOutputStream]
           [java.net URLClassLoader]
           [java.util.concurrent TimeUnit])
  (:gen-class
   :name nihilite.javaagentClasspathDriver
   :prefix "jacd-"
   :main true))

(def pass-count (atom 0))
(def fail-count (atom 0))

(defn pass! [] (swap! pass-count inc))
(defn fail! [] (swap! fail-count inc))

(defn log-cause-chain [t]
  (loop [cause t]
    (when cause
      (println "  " (.getName (class cause))
               (when-let [m (.getMessage cause)] (str ": " m)))
      (recur (.getCause cause)))))

(defn run-nrepl-misc-probe []
  (try
    (let [sys (ClassLoader/getSystemClassLoader)]
      (println "javaagentClasspathDriver: system loader class =" (.getName (class sys)))
      (println "javaagentClasspathDriver: system loader is URLClassLoader?"
               (instance? URLClassLoader sys))
      (println "javaagentClasspathDriver: nrepl/misc.clj via system ="
               (.getResource sys "nrepl/misc.clj"))
      (require 'nrepl.misc)
      (println "javaagentClasspathDriver: require nrepl.misc OK")
      (pass!))
    (catch Throwable t
      (println "javaagentClasspathDriver: require nrepl.misc FAILED")
      (log-cause-chain t)
      (fail!))))

(defn run-require-nrepl-server-probe []
  (try
    (require 'nrepl.server)
    (println "javaagentClasspathDriver: require nrepl.server OK")
    (pass!)
    (catch Throwable t
      (println "javaagentClasspathDriver: require nrepl.server FAILED")
      (log-cause-chain t)
      (fail!))))

(defn run-worker-equivalent-probe []
  (let [worker-error (atom nil)
        worker-done (atom false)
        worker-lock (Object.)
        worker-thread (Thread.
                       (fn []
                         (try
                           (require 'nihilite.registry)
                           (require 'nihilite.boot)
                           (catch Throwable t
                             (reset! worker-error t))
                           (finally
                             (locking worker-lock
                               (reset! worker-done true)
                               (.notifyAll worker-lock)))))
                       "driver-worker-equivalent")]
    (.setDaemon worker-thread true)
    (.start worker-thread)
    (try
      (require 'nihilite.boot)
      (catch Throwable t (log-cause-chain t)))
    (locking worker-lock
      (while (not @worker-done)
        (.wait worker-lock)))
    (if-let [we @worker-error]
      (do
        (println "javaagentClasspathDriver: worker-equivalent concurrent require FAILED")
        (log-cause-chain we)
        (fail!))
      (do
        (println "javaagentClasspathDriver: worker-equivalent concurrent require OK")
        (pass!)))))

(defn run-boot-then-start-server-probe []
  (try
    (require 'nrepl.server)
    (let [handle (boot/start!)]
      (try
        (println "javaagentClasspathDriver: nihilite.boot/start! OK")
        (pass!)
        (finally
          (try (nrepl.server/stop-server handle) (catch Throwable _)))))
    (catch Throwable t
      (println "javaagentClasspathDriver: nihilite.boot/start! FAILED")
      (log-cause-chain t)
      (fail!))))

(defn- read-pipe-into [pipe is buf]
  (let [n (.read is buf)]
    (when (pos? n)
      (locking pipe (.write pipe buf 0 n)))))

(defn- check-markers [state snapshot-string]
  (when (and (not (:init-done? @state)) (.contains snapshot-string "init eval done"))
    (swap! state assoc :init-done? true))
  (when (and (not (:bound? @state)) (.contains snapshot-string "nREPL bencode clients may connect"))
    (swap! state assoc :bound? true)))

(defn spawn-jar-smoke [argv]
  (let [nihilite-jar (nth argv 1)
        init-script (when (> (count argv) 2) (nth argv 2))
        init-form (if (or (nil? init-script) (str/blank? init-script))
                   "(do (require 'clojure.repl) (in-ns 'user))"
                   (str "(load-file \"" (.replace init-script "\\" "\\\\") "\")"))
        cmd ["java"
             "-Djdk.attach.allowAttachSelf=true"
             "-XX:+EnableDynamicAgentLoading"
             (str "-javaagent:" nihilite-jar)
             (str "-Dnihilite.init=" init-form)
             "-Dnihilite.port=0"
             "-Dnihilite.bind=127.0.0.1"
             "-jar" nihilite-jar]
        pb (doto (ProcessBuilder. cmd) (.redirectErrorStream true))
        proc (.start pb)
        is (.getInputStream proc)
        pipe (ByteArrayOutputStream.)
        buf (byte-array 4096)
        reader (Thread.
                (fn []
                  (try
                    (loop []
                      (read-pipe-into pipe is buf)
                      (when (not= -1 (try (.read is buf) (catch Throwable _ -1)))
                        (recur)))
                    (catch Throwable _)))
                "jar-smoke-reader")
        deadline (+ (System/nanoTime) (.toNanos TimeUnit/SECONDS 45))
        state (atom {:bound? false :init-done? false})]
    (.setDaemon reader true)
    (.start reader)
    (loop []
      (when (and (< (System/nanoTime) deadline)
                 (.isAlive proc)
                 (not (and (:init-done? @state) (:bound? @state))))
        (Thread/sleep 200)
        (check-markers state (locking pipe (.toString pipe)))
        (recur)))
    (.destroyForcibly proc)
    (try (.waitFor proc 5 TimeUnit/SECONDS) (catch InterruptedException _ (.interrupt (Thread/currentThread))))
    (let [log (locking pipe (.toString pipe))]
      (if (and (:bound? @state) (:init-done? @state))
        (do
          (println "javaagentClasspathDriver: jar-smoke (nrepl server bound + init ran) OK")
          (pass!))
        (do
          (println "javaagentClasspathDriver: jar-smoke incomplete (bound=" (:bound? @state)
                   ", initDone=" (:init-done? @state) ")")
          (println "...captured log (full):\n" log)
          (fail!))))))

(defn -main [& args]
  (if (and (seq args) (= "spawn-jar-smoke" (first args)))
    (do (spawn-jar-smoke (vec args)) (System/exit 0))
    (do
      (run-nrepl-misc-probe)
      (run-require-nrepl-server-probe)
      (run-worker-equivalent-probe)
      (run-boot-then-start-server-probe)
      (if (and (zero? @fail-count) (= 4 @pass-count))
        (do (println "DRIVER_PASS javaagent classpath: 4/4 probes OK")
            (System/exit 0))
        (do (println "DRIVER_FAIL javaagent classpath: pass=" @pass-count " fail=" @fail-count)
            (System/exit 1))))))
